libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/TimeBasedNextVoiceInstructionsProvider.kt

```diff
@@ -0,0 +1,67 @@
+package com.mapbox.navigation.ui.voice.api
+
+import com.mapbox.api.directions.v5.models.VoiceInstructions
+import com.mapbox.navigation.utils.internal.logW
+
+internal class TimeBasedNextVoiceInstructionsProvider(
+    private val observableTimeSeconds: Int,
+) : NextVoiceInstructionsProvider {
+
+    override fun getNextVoiceInstructions(progress: RouteProgressData): List<VoiceInstructions> {
+        val legs = progress.route.legs() ?: return emptyList<VoiceInstructions>().also {
+            logW(
+                "Route does not contain legs, progress=$progress",
+                LOG_CATEGORY
+            )
+        }
+        var legSteps = legs.getOrNull(progress.legIndex)?.steps()
+        var currentStep = legSteps
+            ?.getOrNull(progress.stepIndex)
+            ?: return emptyList<VoiceInstructions>().also {
+                logW(
+                    "Route does not contain valid step, progress=$progress",
+                    LOG_CATEGORY
+                )
+            }
+
+        val voiceInstructions = mutableListOf<VoiceInstructions>()
+        var cumulatedTime = progress.stepDurationRemaining
+        val currentStepInstructions = currentStep.voiceInstructions()?.filter { instruction ->
+            val distanceAlongGeometry = instruction.distanceAlongGeometry()
+            distanceAlongGeometry != null
+                && distanceAlongGeometry <= progress.stepDistanceRemaining
+        }
+        if (currentStepInstructions != null) {
+            voiceInstructions.addAll(currentStepInstructions)
+        }
```

Comment:
Maybe extract a function `getNotPassedVoiceInstuctions(step, progress.distanceRemaining)`?

Reply:
`distanceRemaining` is not used for any steps other than current. For the next steps the logic is a bit different. So the method would not be reused. Why extract then? I think it would only complicate things.

Reply:
> Why extract then? 

functions's names could be used as a documentation. Reading function's name, a programmer can understand what to expect in the code. It also improve understanding when you can observe algorithms from different levels. For example, here's how it can be represented on the hight level:
```kotlin
val result = findNotPassedVoiceInstructionsFromCurrentStep(...) + findStepsAheadInTime(observableTimeSeconds).map { it.voiceInstruction }
```
And then, if you want to check specific details, you can dive into the functions to see how exactly it finds instructions from the current step, or how it iterates over the steps

Reply:
I'll try to think of something.